### DISCUSSION TITLE
Refactor "type" into something else

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/indexers#2] [General]
This is how we are currently defining the `ContractType`, `StreamType`, and the `ActionType`:

- https://github.com/sablierhq/v2-subgraphs/blob/b066883b2dacd26acd91ba6b73b4b897a4554ebe/packages/core/schema.graphql#L26
- https://github.com/sablierhq/v2-subgraphs/blob/b066883b2dacd26acd91ba6b73b4b897a4554ebe/packages/core/schema.graphql#L41
- https://github.com/sablierhq/v2-subgraphs/blob/b066883b2dacd26acd91ba6b73b4b897a4554ebe/packages/core/schema.graphql#L88

My concern is that this key can be easily mingled with the reserved `type` keyword in GraphQL - I am surprised that GraphQL even allows this word to be used as a key within an actual type.

Proposed alternatives to clear up any doubt:

- `class`
- `category`
- `entityType`

**Comment by razgraf** [comment]
I agree that it might be better to rename it, although not because of GraphQL. 

GraphQL or Typescript using that "type" keyword will not create any difficulties whatsoever. 
1. One is used as an entity field, the other in the higher context defining a new structure shape. A language interpreter will catch this easily.
2. Avoiding "type" in free speech is quite impossible, which translates to the word being used a lot in code.

Having said that, I'm not the biggest fan of "type" either. My reasoning is not just reserved keywords but 
1. it being already used in other libraries. This can cause internal fields to be overridden, with the best example being xState machines (dispatch will include a type field).
2. An entity having multiple ... types of `type`. Whenever an item has 2 characteristics with different value sets, it can be described as having 2 types. Picking one makes it tough to work around it when extending the entity definition.

The **v2-interfaces** already implement a workaround for `entity.type` by using words such as `context` or `purpose`.

From your proposal, `category` sounds good. 
For `class`, it would suffer from the same contextual usage as `type` (Even more due to it being a JS word).
For `entityType` it's quite repetitive, as well as incorrect. The entity type of a variable is the `type X @entity` itself (X).

**Reply by PaulRBerg** [reply]
Great analysis, agree with everything. `category` is good.
